Expose csi (as metadata)

antonott opened the pull request:
So this is quite similar to an [earlier pull request](https://github.com/bastibl/gr-ieee802-11/pull/225) that seems to have gone stale, but my approach is a little bit different. While the idea there was to expose CSI as an output stream from the Frame Equalizer block, my approach is instead to put the CSI as metadata in the tag stream associated with the regular output of the Frame Equalizer. Since there has been interest before I thought I might write about it here instead of just working on it alone in my fork of the repo.

The main benefit of this approach is that we get the association between a particular frame and its CSI-vector for free, because the stream tags are already passed on a per frame basis. If, one the other hand, the CSI was exposed as an output stream we would lose that association, and would have to reconstruct it somehow if it was important to our application. This seems likely with many applications, including mine. With two different output streams that could be quite complicated.

Another benefit, in my view, is that it puts the CSI on equal footing with the other frame metadata, which I think is appropriate on a conceptual level. Its also easy to then implement sending that data along, say from the Decode MAC block to the Parse MAC block.

(Oh, and I added "beta" to the metadata as well. Might be useful in some applications, and I think it conceptually qualifies as metadata.)

antonott committed:
Make `d_H` inherited from base class.

All equalizer classes had their own private `d_H` vector, despite all being derived from a common `base` class. Now `d_H` is instead a protected member of `base`, and thus accessible to the derived equalizer classes.

antonott committed:
Add base::get_csi() to expose d_H.

antonott committed:
Skip guard bands and null subcarrier in get_csi(). Add beta and csi metadata to stream tag from frame equalizer.

antonott committed:
Add beta and csi to decode_mac PDU output.

Also changed to using `std::vector<gr_complex>` type because it was becoming more and more inconvenient to use c-style arrays.

antonott commented:
I have also been looking at extending the Parse MAC block. As it is now it decodes the MAC frame and, if debug is enabled, prints the decoded information to the console. If debug is disabled it does almost nothing (just the FER estimate). Would it not make sense to send the decoded data along instead? Something like this, where the parentheses are [cons](https://en.wikipedia.org/wiki/Cons) as usual:
```
                             ┌─────────┐
(dict meta, blob msg)-------►│Parse MAC├-------►(dict meta, dict msg)
                             └─────────┘
```
The outgoing msg dictionary would have `key : value` entries such as, say, `'frame_control' : 'DATA'`, `'seq_num' : 54`, `'data' : 'Hello world'` and so on. Alternatively one might do something like
```
-------►(dict meta, string data)
```
for the output, with only the decoded `data` field in the `cdr` and all of the other decoded fields entered into the metadata dictionary. (I suppose the `string` type would actually be `pmt::symbol` or some kind of vector.)

Either way, it would be easy to design and append application specific blocks that use the frame data without having to reimplement the parsing. Some examples might be blocks that filter on certain fields (say MAC addresses), blocks that print some of the data and/or metadata to a file, etc.

As for the FER output that is already present, I would advocate for putting that in the metadata dictionary as well.

bastibl commented:
Great, very happy you pick up this work. The request to add CSI (or rather the question how to get it out of the receiver) comes up again and again.

I agree that adding the CSI as metadata makes sense. An advantage of the other approach was that you can easily connect a graphical sink an watch the CSI. Maybe, there is a way to do it with GNU Radio Message tools? Or one could consider extending, for example, the Parse MAC block to write CSI to a stream output. At this stage, one could even filter for a particular receiver, which would be cool. Or a dedicated separate block?

Not sure about the other extension of the Parse MAC block. It is a very, very limited block that I only used for basic debugging. If one is interested in the payload, I'd recommend to use Wireshark (or Scapy). But, at the same time, if you find this helpful, I'm happy to merge it. It doesn't increase complexity, so why not.

You marked this as  a Draft. Let me know, if I should have a more detailed look and test it.

antonott commented:
> I agree that adding the CSI as metadata makes sense. An advantage of the other approach was that you can easily connect a graphical sink an watch the CSI. Maybe, there is a way to do it with GNU Radio Message tools? Or one could consider extending, for example, the Parse MAC block to write CSI to a stream output. At this stage, one could even filter for a particular receiver, which would be cool. Or a dedicated separate block?

I think a dedicated block that picks out the CSI would be preferable. Then it can be placed anywhere in the Frame Equalizer, Decode MAC, Parse MAC chain. Consider this:
```
                   ┌───────────┐
(meta, message)---►│Extract CSI├---►(meta, csi)
                   └───────────┘
```
With `csi` being of a uniform vector type, and `meta` being a dictionary, the outgoing message will have the format of a [GNU Radio PDU](https://wiki.gnuradio.org/index.php/Message_Passing#PDUs). That means one could very easily do things like this:
```
                   ┌───────────┐                   ┌────────────────────┐    ┌─────────────────────┐
(meta, message)---►│Extract CSI├---►(meta, csi)---►│PDU to Tagged Stream├───►│QT GUI Frequency Sink│
                   └───────────┘                   └────────────────────┘    └─────────────────────┘
```
Very flexible!

> Not sure about the other extension of the Parse MAC block. It is a very, very limited block that I only used for basic debugging. If one is interested in the payload, I'd recommend to use Wireshark (or Scapy). But, at the same time, if you find this helpful, I'm happy to merge it. It doesn't increase complexity, so why not.

In my application I will be filtering on MAC address and then feeding the CSI (and possibly other metadata) to an external python program. Maybe that is easier to do with Wireshark and Scapy - I am not very familiar - but that is why I wanted to access some of the decoded fields. I might implement this in a different branch, so that you can make a separate decision on whether or not you want to merge it in the end.

bastibl commented:
One could extend the Parse MAC block as you suggested, then use the PDU Filter block of GNU Radio to extract only a single sender, for example, and then use the CSI extractor to plot it interactively. I'd just directly convert from PDU to stream. The trick with converting the metadata of a PDU to a PDU of the metadata will probably confuse people.

Also, I completely agree that processing everything in an external script is the way to go. But it would still be nice to have an interactive example.

What do you think about this?

antonott commented:
> One could extend the Parse MAC block as you suggested, then use the PDU Filter block of GNU Radio to extract only a single sender, for example, and then use the CSI extractor to plot it interactively.

To be able to use the PDU Filter block like that we must put the decoded fields in the meta part of the PDU (see [documentation](https://wiki.gnuradio.org/index.php/PDU_Filter)) and the cdr must be a uniform vector type. So I think the logical choice would be to put the decoded data field of the MAC frame in the cdr. That is,
```
               ┌─────────┐                     ┌──────────┐
(meta, msg)---►│Parse MAC├---►(meta', data)---►│PDU Filter├---►
               └─────────┘                     └──────────┘
```
with the outgoing `meta'` containing all the other decoded fields, in addition to what was already in the meta dictionary (CSI etc.). Does that make sense?

> I'd just directly convert from PDU to stream. The trick with converting the metadata of a PDU to a PDU of the metadata will probably confuse people.

I suppose that might be better. First I thought it would be less flexible, but if we also propagate the metadata in a stream tag one could always use a [Tagged Stream to PDU](https://wiki.gnuradio.org/index.php/Tagged_Stream_to_PDU) block to convert it to the format from my approach. Hence both approaches are functionally equivalent, but yours is probably the more user friendly one.

bastibl commented:
> To be able to use the PDU Filter block like that we must put the decoded fields in the meta part of the PDU (see documentation) and the cdr must be a uniform vector type. So I think the logical choice would be to put the decoded data field of the MAC frame in the cdr. That is,
>
> with the outgoing meta' containing all the other decoded fields, in addition to what was already in the meta dictionary (CSI etc.). Does that make sense?

Makes sense to me :-) And should be inline with your proposal to extend Parse MAC block, no? Only, I wouldn't touch the payload (`msg == data`) and only add decoded fields as additional metadata to the dictionary of the PDU.

antonott commented:
> Only, I wouldn't touch the payload (`msg == data`) and only add decoded fields as additional metadata to the dictionary of the PDU.

After thinking about it I agree.

On a more technical note, what is going on [here](https://github.com/bastibl/gr-ieee802-11/blob/d8f45138a232ee54dd22be8c14c02412fb8a12de/lib/parse_mac.cc#L145-L146)? It seems strange to call a pointer `len`, and even stranger to compare an integer `length` with a pointer. I might be missing something though, as I don't have much experience with C++.

antonott committed:
Parse MAC writes to meta and propagates PDU.

This commit extends the Parse MAC block so that it puts the parsed fields in the metadata dictionary of the incoming PDU and also propagates it on a message output. It does not modify the cdr of the PDU. The 'fer' output has been removed, and instead 'Lost frames' and 'Instantanteous FER' are put in the metadata dictionary.

antonott committed:
Small fix to last commit

If we cannot determine the length of the mac frame body, we should probably not attempt to parse it.

antonott commented:
The two latest commits implement the extension to the Parse MAC block that we have been discussing. I verified that it is possible to feed the output through a PDU Filter block to filter on MAC addresses.

It is not the most elegant implementation, but it seems to do the job. If at some point someone decides to "flesh out" the Parse MAC block a more drastic rewrite might be in order.

bastibl commented:
> On a more technical note, what is going on here? It seems strange to call a pointer len, and even stranger to compare an integer length with a pointer. I might be missing something though, as I don't have much experience with C++.

`length` is the length of the buffer, i.e., the payload of the frame. The SSID of the network (the name shown in the list of networks) is encoded as type-length-value in a beacon frame. Its length (i.e., the number of characters of the name) is encoded at byte offset 38. The code makes sure that the frame actually has a size of at least 38 and creates `len`, a pointer to that offset. Then, it does more checks and creates a string with the corresponding size, which includes the name of the network. It never adds the `len` pointer but the value it is pointing to.

bastibl commented on lib/parse_mac.cc:
```diff
@@ -37,69 +37,86 @@ parse_mac_impl(bool log, bool debug) :
+
+	int body_len = frame_len - body_start;
+
+	std::string frame_body(frame + body_start, body_len);
```
I don't think this is a good idea. The frame body is binary data and I'd just leave it as that (i.e., restore print ascii for debug output and not add this as field).

antonott replied:
Fair enough, I can revert that change.

bastibl commented on lib/parse_mac.cc:
```diff
@@ -280,87 +344,78 @@ void parse_control(char *buf, int length) {
-
-}
-
-void print_mac_address(uint8_t *addr, bool new_line = false) {
```
Afais, there is not replacement for this function. `format_mac` is incomplete.

antonott replied:
`print_mac_address` could not be used when inserting MAC addresses into the meta dictionary, only to print to console, so I replaced it with `format_mac` which does only the formatting.

But I see now that removing the `setfill` and `setw` might give rise to issues, so I will fix that. Did you have something else in mind?

bastibl replied:
I thought there would be more missing. But yes, formatting would be nice. I think you can use a string stream for more direct refactoring.

bastibl reviewed (changes requested):
Looks very good, except for the two minor comments. Would you also add a block to filter and output CSI and add it to the example flowgraphs?

antonott committed:
Changes in response to discussion with bastibl.

We will not parse the frame body and put it in the metadata. Instead `print_ascii` has been restored for debug output. `format_mac` has been made more complete, with correct hex formatting.

antonott commented:
> Would you also add a block to filter and output CSI and add it to the example flowgraphs?

Sure, but you don't mean a single block right? You mean something like this, as we discussed earlier?
```
    ┌─────────┐    ┌──────────┐    ┌───────────┐    ┌─────────────────────┐
---►│Parse MAC├---►│PDU Filter├---►│Extract CSI├───►│QT GUI Frequency Sink│
    └─────────┘    └──────────┘    └───────────┘    └─────────────────────┘
```
I just need to implement the Extract CSI block first.

antonott commented:
I have been trying to implement an Extract CSI block, closely based on the PDU to Tagged Stream block, but it does not work. You can see my attempt [here](https://github.com/antonott/gr-ieee802-11/compare/expose_csi...antonott:csi_block_as_stream_block) if you are interested. The issue is that the `calculate_output_stream_length` function, which all blocks derived from `tagged_stream_block` must override, is never called. After fighting with it for a few hours I stumbled over [this guide](https://wiki.gnuradio.org/index.php/Tagged_Stream_Blocks) which shows the central role of the "length tag"  for tagged stream blocks. Since a length tag doesn't make much sense for us I had just ignored it, but it seems like that may be why I couldn't get it to work.

I think it might actually make more sense to implement this as a source block. USRP source blocks need to wait for data to come from the USRP, just like this block will have to wait for incoming PDUs. So I think if I write Extract CSI as a source block, the scheduler should be able to handle it.

bastibl commented:
I agree that you don't have to use a Tagged Stream Block. The issue with your code is that
- you have to give the tag a name, not ""
- "pdu in" vs "pdu int"
- you overwrite and, therefore, shadow the class variable `d_meta` in `calculate_output_stream_length`
- the second check (`(unsigned int)noutput_items >= 52`) should be removed in the assert, since your "item" is the vector of 52 samples.

One idea: instead of `d_csi_ready`, you could check if the vector is `empty()` and `clear()` it after copying.

antonott commented:
Well spotted! I guess I got a bit of tunnel vision working on this for so long. After fixing those issues `calculate_output_stream_length` is indeed called correctly, the latter parts of the `work` function are being executed, and so on. But I still get no output when connecting Extract CSI -> Complex to Mag -> QT GUI Time Sink, so it seems there is something fishy with the scheduler interaction.

But as I said earlier, I will try to implement it in the style of a source block instead. Our desired behaviour doesn't really seem to fit that well in the `tagged_stream_block` mold.

bastibl commented:
I tried the Vector Sink for the GUI, but it didn't work. (Maybe an upstream issue).
For Time Sink, you need to convert Vector to Stream, set the number of points to 52, and zoom out (or toggle auto scale).

antonott commented:
"Time Sink" in my previous comment was a typo. I meant "QT GUI Vector Sink". I will try with Time Sink the way you say, for debug.

antonott commented:
Well, that seems to work, at least! Still, I don't like it, because it is confusing for the user. Not least that the Time Sink has a Sample rate parameter, something that we don't otherwise need in the wifi_loopback example.

I did some very elementary tests with the QT GUI Vector Sink to see if it is completely broken, but it seems to work fine with other blocks. Again, I will try to reimplement the Extract CSI block as a source to see if that resolves it.

antonott committed:
Added Extract CSI block.

This block takes PDU messages as input, looks for a "csi" entry in the meta dictionary, and if found outputs a complex CSI vector on a stream output. The meta is also sent along as stream tags.

antonott commented:
The latest commit implements a simpler Extract CSI block that works more like a typical source block. It works with the -> Vector to Stream -> QT GUI Time Sink method, but just like the previous implementation it does not work with the QT GUI Vector Sink. I am starting to agree that it might be an upstream issue.

I don't think I can motivate spending more time troubleshooting with the vector sink at the moment, so I suggest going forward with the other method of visualization. Does that sound fine with you? The only really anoying thing is, again, that we need to provide a dummy sample rate to the time sink, but maybe we can live with that.

By the way, is the name "Extract CSI" okay, or should we call it "WiFi Extract CSI" for consistency?

bastibl commented:
I merged it with some follow-up changes. Thank you very much! Very happy that this is now supported.

antonott commented:
Great! Happy to contribute!
